File: Bismark2Matrix.R
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

# Rscript --vanilla script.R file_path file_tsv file_out
# Reproducible PCA-based approach for methylation

# Combine bismark.cov files into a matrix
# Here the script checks for headers, and only proceeds if the first column of the file_tsv input contains files with a ".cov" 
covList <- read.delim(args[2], header=F)
if(grepl("\\.cov", covList[1,1]) == F){
  covList <- read.delim(args[2], header=T)
}
if(grepl("\\.cov", covList[1,1])==F){
  message(paste("Error:", args[2], "column 1 does not contain .cov files!")) # break the script if there are no .cov files supplied
} else{
  for(i in 1:nrow(covList)){
    name <- as.character(covList[i,2])
    condition <- as.character(covList[i,3])
    message(paste0("processing ", name, "..."))
    df <- read.delim(paste0(args[1],"/",as.character(covList[i,1])), header=F)
    df$cpgID <- paste0(df$V1,":",df$V2)
    df$count <- df$V5+df$V6
    df <- df[c(7,4,8)]
    colnames(df) <- c("cpgID", paste0(name, "_", condition, "_PercMeth"), paste0(name, "_", condition, "_nCpG"))

    if(i == 1){
      output <- df
    } else {
      message(paste("merging", name, "to output..."))
      output <- merge(output, df, by="cpgID", all=F) # remove NA rows
    }
  }
  message("Removing NA rows...")
  output <- na.omit(output)
  message("Writing output...")
  write.table(output, args[3], sep="\t", quote=F, col.names = T, row.names = F)
}

File: Bismark2Matrix.py
import pandas as pd
from functools import reduce
import sys, getopt

# Command Line Inputs:
def main(argv):
   file_tsv = ''
   file_out = ''
   file_path = ''
   
   try:
      opts, args = getopt.getopt(argv, "hi:o:p:", ["file_tsv=","file_out=","file_path="])
   except getopt.GetoptError:
      print('\nBismark2Matrix.py\n-i <tab-delimited input sample file (file name, sample name, treatment group); expects column names in row 1>\n-o <output file name>\n-p <path to input files>\n')
      sys.exit(2)
   
   for opt, arg in opts:
      if opt == '-h':
         print('\n--- Bismark2Matrix Required Inputs ---\n\n\t-i | --file_tsv\t\t<tab-delimited input sample file (filename, sample, group)>\n\t-o | --file_out\t\t<output file name>\n\t-p | --file_path\t<path to input files>\n\n')
         sys.exit()
      elif opt in ("-i", "--file_tsv"):
         samples = arg
      elif opt in ("-o", "--file_out"):
         outPath = arg
      elif opt in ("-p", "--file_path"):
         inPath = arg
   
   dfList =[]
   sample_file = pd.read_csv(samples, sep="\t",  low_memory=False)
   for i, row in sample_file.iterrows():
       # get info and read in cov file for one sample:
       sample_input=str(sample_file.iat[i, 0])
       sample_name=str(sample_file.iat[i, 1])
       sample_group=str(sample_file.iat[i, 2])
       print("processing "+sample_name+"...")
       cov = pd.read_csv(inPath+'/'+sample_input, sep="\t",  low_memory=False, header=None)
       
       # add cpgID to first columns
       ident = cov[0] + ":" + cov[1].apply(str)
       cov.insert(0, "cpgID", ident, True)
       # add %methylation and nCPGs to the second third columns:
       count = cov[4] + cov[5]
       cov = cov.loc[:, ["cpgID",3]]
       cov.insert(2, sample_name+"_"+sample_group+"_nCpG", count, True)
       cov.rename(columns={3:sample_name+"_"+sample_group+"_PercMeth"}, inplace=True)
       dfList.append(cov)
       
   # Combine all the loci on cpgID, removes NA sites
   print("merging samples...")
   out=reduce(lambda x, y: pd.merge(x, y, on = 'cpgID'), dfList)
   print("Writing output...")
   out.to_csv(outPath, sep="\t", index=False) 
   print("Done :)")
   
if __name__ == "__main__":
   main(sys.argv[1:])






